include/opw_kinematics/opw_kinematics_impl.h

```diff
@@ -132,69 +133,25 @@ Solutions<T> inverse(const Parameters<T>& params, const Transform<T>& pose) noex
-  out2[4] = (theta5_iii + params.offsets[4]) * params.sign_corrections[4];
-  out2[5] = (theta6_iii + params.offsets[5]) * params.sign_corrections[5];
-
-  Solution<T>& out3 = solutions[3];
-  out3[0] = (theta1_ii + params.offsets[0]) * params.sign_corrections[0];
-  out3[1] = (theta2_iv + params.offsets[1]) * params.sign_corrections[1];
-  out3[2] = (theta3_iv + params.offsets[2]) * params.sign_corrections[2];
-  out3[3] = (theta4_iv + params.offsets[3]) * params.sign_corrections[3];
-  out3[4] = (theta5_iv + params.offsets[4]) * params.sign_corrections[4];
-  out3[5] = (theta6_iv + params.offsets[5]) * params.sign_corrections[5];
-
-  Solution<T>& out4 = solutions[4];
-  out4[0] = (theta1_i + params.offsets[0]) * params.sign_corrections[0];
-  out4[1] = (theta2_i + params.offsets[1]) * params.sign_corrections[1];
-  out4[2] = (theta3_i + params.offsets[2]) * params.sign_corrections[2];
-  out4[3] = (theta4_v + params.offsets[3]) * params.sign_corrections[3];
-  out4[4] = (theta5_v + params.offsets[4]) * params.sign_corrections[4];
-  out4[5] = (theta6_v + params.offsets[5]) * params.sign_corrections[5];
-
-  Solution<T>& out5 = solutions[5];
-  out5[0] = (theta1_i + params.offsets[0]) * params.sign_corrections[0];
-  out5[1] = (theta2_ii + params.offsets[1]) * params.sign_corrections[1];
-  out5[2] = (theta3_ii + params.offsets[2]) * params.sign_corrections[2];
-  out5[3] = (theta4_vi + params.offsets[3]) * params.sign_corrections[3];
-  out5[4] = (theta5_vi + params.offsets[4]) * params.sign_corrections[4];
-  out5[5] = (theta6_vi + params.offsets[5]) * params.sign_corrections[5];
-
-  Solution<T>& out6 = solutions[6];
-  out6[0] = (theta1_ii + params.offsets[0]) * params.sign_corrections[0];
-  out6[1] = (theta2_iii + params.offsets[1]) * params.sign_corrections[1];
-  out6[2] = (theta3_iii + params.offsets[2]) * params.sign_corrections[2];
-  out6[3] = (theta4_vii + params.offsets[3]) * params.sign_corrections[3];
-  out6[4] = (theta5_vii + params.offsets[4]) * params.sign_corrections[4];
-  out6[5] = (theta6_vii + params.offsets[5]) * params.sign_corrections[5];
-
-  Solution<T>& out7 = solutions[7];
-  out7[0] = (theta1_ii + params.offsets[0]) * params.sign_corrections[0];
-  out7[1] = (theta2_iv + params.offsets[1]) * params.sign_corrections[1];
-  out7[2] = (theta3_iv + params.offsets[2]) * params.sign_corrections[2];
-  out7[3] = (theta4_viii + params.offsets[3]) * params.sign_corrections[3];
-  out7[4] = (theta5_viii + params.offsets[4]) * params.sign_corrections[4];
-  out7[5] = (theta6_viii + params.offsets[5]) * params.sign_corrections[5];
+  Eigen::Map<const Eigen::Array<T, 6, 1>> offsets(params.offsets.data());
+  Eigen::Array<T, 6, 1> signs;
+  signs[0] = params.sign_corrections[0];
+  signs[1] = params.sign_corrections[1];
+  signs[2] = params.sign_corrections[2];
+  signs[3] = params.sign_corrections[3];
+  signs[4] = params.sign_corrections[4];
+  signs[5] = params.sign_corrections[5];
```

Comment:
Could you avoid this with the following?:

```c++
Eigen::Map<const Eigen::Array<signed char, 6, 1>> signs(params.sign_corrections);

...

output = (theta.colwise() + offsets).colwise() * signs.cast<T>();
```

Reply:
Yea, I originally tried that and Eigen does not like the `signed char` type for some reason.

Reply:
Bummer; never mind then.